### PR TITLE
[RovingFocusGroup] Refactor internals to take into account hidden items

### LINKED ISSUE #581: [DropdownMenu] Items that are invisible cause focus issues
### Current Behavior

When navigating with the arrow keys through a DropdownMenu that has hidden items (ie: with `display: none`), the focus selection stops when reaching the hidden item, and as such doesn't focus all items.

### Expected behavior

The visible items are focused when pressing the arrow keys, even if there are hidden items.

### Reproducible example

1. Open the Dropdown
2. Press the arrow down key repeatedly
3. Verify that the last element doesn't get focused

[CodeSandbox Example](https://codesandbox.io/s/relaxed-mestorf-ru0rd?file=/src/App.tsx)

### Suggested solution

Filter out the invisible items. I already have a PR ready for it (that includes a Storybook example).

### Additional context

N/A

### Your environment

| Software         | Name(s) | Version |
| ---------------- | ------- | ------- |
| Radix Package(s) |   `@radix-ui/react-dropdown-menu`      |   0.0.17    |
| React            | n/a     |   17      |
| Browser          |         |     Brave    |
| Assistive tech   |         |         |
| Node             | n/a     |     14    |
| npm/yarn         |         |     yarn 1.x    |
| Operating System |         |    MacOS     |

### PR DESCRIPTION
**Pull request by benoitgrelard**
Closes #581 

---
> Note: Better view with whitespace turned off
---

This PR primarily refactors the internals of `RovingFocusGroup` so we can take into account the fact that hidden items don't get focused by the browser.

The crux of the implementation here is that we've introduced a wrapper as part of that API which will act as a proxy to send focus programatically to the correct place. We have a reference to all items in the group and so, we can programatically and progressively attempt to focus the "first" item, and eventually land on the correct one (the first visible one) without having to resort to calculating computed styles to know if an item is indeed hidden.

Once that was done, all of the dependent packages needed updated accordingly.

This has made the overall implementation(s) more declarative, and more the React way.

> Note: There are now technically new changes to document for some packages because they now get the underlying props from `RovingFocusGroup`
- ✨ `RadioGroup` now gets new `orientation`, `dir` and `loop` props
- ✨ `ToggleGroup` now gets new `orientation`, `dir` and `loop` props 

cc @andy-hook, may be good for you to take a quick look through this to see if there's any impact on what you've been working on.

### CODE REVIEW

File: packages/react/menu/src/Menu.tsx
Comment (benoitgrelard) [inline packages/react/menu/src/Menu.tsx:664]: This is duplicated from `RovingFocusGroup`, let me know if you think we should do something about that.

File: packages/react/menu/src/Menu.tsx
Comment (benoitgrelard) [reply packages/react/menu/src/Menu.tsx:386]: Also, this is analogous (duplicated) to what we do in `RovingFocusGroup`. Let me know what you think.

File: packages/react/radio-group/src/RadioGroup.tsx
Comment (benoitgrelard) [inline packages/react/radio-group/src/RadioGroup.tsx:26]: Removed all the dead code about handling no-roving case as it was a leftover from its use in `Toolbar` which has now been replaced by `ToggleGroup`.

File: packages/react/radio-group/src/RadioGroup.tsx
Comment (benoitgrelard) [inline packages/react/radio-group/src/RadioGroup.tsx:61]: These are now new props for `RadioGroup` and will need documenting.

File: packages/react/radio-group/src/RadioGroup.tsx
Comment (benoitgrelard) [inline packages/react/radio-group/src/RadioGroup.tsx:110]: Fixed the remove `name` prop as it's a DOM attribute it would get re-added otherwise.

File: packages/react/roving-focus/src/RovingFocusGroup.tsx
Comment (benoitgrelard) [inline packages/react/roving-focus/src/RovingFocusGroup.tsx:124]: Like I said, not a huge fan of the name here, would gladly take other suggestions :)

File: packages/react/roving-focus/src/RovingFocusGroup.tsx
Comment (benoitgrelard) [inline packages/react/roving-focus/src/RovingFocusGroup.tsx:277]: This is duplicated from `menuTypeahead.tsx`. Let me know if you think we should do something about it.

File: packages/react/toggle-group/src/ToggleGroup.tsx
Comment (benoitgrelard) [inline packages/react/toggle-group/src/ToggleGroup.tsx:212]: These are new props for `ToggleGroup` which will need documenting.

File: packages/react/toggle-group/src/ToggleGroup.tsx
Comment (benoitgrelard) [inline packages/react/toggle-group/src/ToggleGroup.tsx:256]: I changed this to calculate `pressed` and `disabled` in one place, and pass them down to both implementation.

File: packages/react/roving-focus/src/RovingFocusGroup.tsx
Comment (jjenzz) [inline packages/react/roving-focus/src/RovingFocusGroup.tsx:173]: Do we still need this ref? I can't see it being used.

File: packages/react/tabs/src/Tabs.tsx
Comment (jjenzz) [inline packages/react/tabs/src/Tabs.tsx:168]: Do we want to move these below `...tabProps` like we did with `name` in `RadioGroup` because we are omitting them from the types?

File: packages/react/toggle-group/src/ToggleGroup.tsx
Comment (jjenzz) [inline packages/react/toggle-group/src/ToggleGroup.tsx:257]: We've avoided naming these sorts of conditionals `Comp` in the past I believe because it clashes with when we do `as: Comp = DEFAULT_TAG`. 

For a second I thought that this wasn't using `Primitive` but it is.

File: packages/react/toolbar/src/Toolbar.tsx
Comment (jjenzz) [inline packages/react/toolbar/src/Toolbar.tsx:24]: It's not rendering `Primitive` anymore so `typeof TOOLBAR_DEFAULT_TAG`

**Review by jjenzz — CHANGES_REQUESTED** [review]
Super excited about this :tada:

I noticed one little thing when testing that felt a bit odd to me. If the items have been interacted with, and then the focused one is hidden/removed, it feels like we need to reset `hasInteracted`.

CleanShot 2021-04-26 at 12 34 44

File: packages/react/menu/src/Menu.tsx
Comment (benoitgrelard) [reply packages/react/menu/src/Menu.tsx:418]: It's not always on mouse leave though, as you'll see below, the same callback is called when mouse moves over a disabled item for example.

File: packages/react/roving-focus/src/RovingFocusGroup.tsx
Comment (benoitgrelard) [reply packages/react/roving-focus/src/RovingFocusGroup.tsx:173]: Great catch, no needed anymore, forgot to delete when we removed all the other code.

File: packages/react/tabs/src/Tabs.tsx
Comment (benoitgrelard) [reply packages/react/tabs/src/Tabs.tsx:168]: Yeah, can do. Do you think we'd need a similar comment?

File: packages/react/toggle-group/src/ToggleGroup.tsx
Comment (benoitgrelard) [reply packages/react/toggle-group/src/ToggleGroup.tsx:257]: Would you rather I keep the conditional inline and duplicate the ref/pressed/disabled props?

File: packages/react/toolbar/src/Toolbar.tsx
Comment (benoitgrelard) [reply packages/react/toolbar/src/Toolbar.tsx:24]: Yep, missed that one!

File: packages/react/menu/src/Menu.tsx
Comment (benoitgrelard) [reply packages/react/menu/src/Menu.tsx:68]: I haven't no, it's just something that came to my mind when I was trying out that sandbox. I'll dig.

File: packages/react/menu/src/Menu.tsx
Comment (benoitgrelard) [reply packages/react/menu/src/Menu.tsx:116]: I'm not sure I follow here. We still need to "declare" the props this component takes no?

File: packages/react/menu/src/Menu.tsx
Comment (benoitgrelard) [reply packages/react/menu/src/Menu.tsx:293]: We have that name in `FocusScope` yes, but that's because it's truly on mount.
This here, doesn't have anything to do with mount or unmount, it's the refocus that happens when the wrapper gets focused.

File: packages/react/menu/src/Menu.tsx
Comment (benoitgrelard) [reply packages/react/menu/src/Menu.tsx:386]: Great. I agree btw 🙂

**Comment by benoitgrelard** [comment]
> Super excited about this 🎉
> 
> I noticed one little thing when testing that felt a bit odd to me. If the items have been interacted with, and then the focused one is hidden/removed, it feels like we need to reset `hasInteracted`.
> 
> CleanShot 2021-04-26 at 12 34 44
> 
> Happy to have a call about this one if i'm not making sense 🙂

What I see in the gif seems to make sense to me, so yeah perhaps we should have a call on that particular one!

File: packages/react/menu/src/Menu.tsx
Comment (jjenzz) [reply packages/react/menu/src/Menu.tsx:116]: Yeah, but we only extend the `OwnProps` like this when the props are being spread onto that element, which they aren't here. 

You can see below in this file you have done `onEscapeKeyDown?: DismissableLayerProps['onEscapeKeyDown'];` to pull out the props from wrapping components but you haven't merged all their props in.

Might be easier to call on this one?

File: packages/react/menu/src/Menu.tsx
Comment (jjenzz) [reply packages/react/menu/src/Menu.tsx:293]: So is it just `onFocusEnter` like `onMouseEnter` ? When focus enters the roving focus group we want to prevent it?

File: packages/react/menu/src/Menu.tsx
Comment (jjenzz) [reply packages/react/menu/src/Menu.tsx:418]: I'm confused by that one. I'm thinking that if a `disabled` item is hovered, we either have no items hovered yet or an item's mouse leave was triggered. I think I need you to help me understand this one on our call but either way, we can have another callback for that if needed. My point here is more about changing this to events and not commands.

File: packages/react/tabs/src/Tabs.tsx
Comment (jjenzz) [reply packages/react/tabs/src/Tabs.tsx:168]: I don't think so personally, because i think that when things are moved below props spread, they are there for a reason. But I know you like to keep things consistently ordered so might be more of an issue for you looking at your own code later?

File: packages/react/toggle-group/src/ToggleGroup.tsx
Comment (jjenzz) [reply packages/react/toggle-group/src/ToggleGroup.tsx:257]: Yeah I'm okay with that duplication in this instance because it's super simple, nothing complex going on there.

File: packages/react/roving-focus/src/RovingFocusGroup.tsx
Comment (andy-hook) [inline packages/react/roving-focus/src/RovingFocusGroup.tsx:206]: I have a case where I'd like to handle `keydown` in a parent element rather than binding directly to the roving items, has this come up before? wondering whether it's worth having an option for this.

Comment (benoitgrelard) [reply packages/react/roving-focus/src/RovingFocusGroup.tsx:206]: You mean perhaps it's not a good idea to stop propagation early here?

This made me realise that it's not needed anymore actually because I now check on the line above that it's an even happening on the current target, and not a bubbling event.

So I can remove it, and I think that solves your question too then?

File: packages/react/menu/src/Menu.tsx
Comment (benoitgrelard) [reply packages/react/menu/src/Menu.tsx:68]: Heads up, after checking this intuition, it turned out to be true, and if the users memoizes an item, it absolutely destroys our solution and makes the order unpredictable… So we're back to the drawing board on the tracking aspect. After discussing with @jjenzz, we think we're going to go back to the DOM again but perhaps in something a bit more of a general solution as it's needed in a few places, and we realised that other components are broken because of the same issue.

File: packages/react/roving-focus/src/RovingFocusGroup.tsx
Comment (andy-hook) [reply packages/react/roving-focus/src/RovingFocusGroup.tsx]: Yes exactly, this solves it.

File: packages/react/menu/src/Menu.tsx
Comment (benoitgrelard) [reply packages/react/menu/src/Menu.tsx:296]: Well that's the thing, `onFocusEnter` just sounds like `onFocus`, which is the event we hook onto. But because this is a custom event, I'd rather want to name it according to what it does, rather than what event it hooks on so.
If I describe what the event does (with its associated behaviour that you can prevent) is:

> when the wrapper itself gets focused, move focus to intended item

That's what the name should be based on.

File: packages/react/collection/src/Collection.tsx
Comment (benoitgrelard) [inline packages/react/collection/src/Collection.tsx:53]: had to do `as unknow as` for some reason once I added `React.forwardRef`

File: packages/react/roving-focus/src/RovingFocusGroup.tsx
Comment (benoitgrelard) [inline packages/react/roving-focus/src/RovingFocusGroup.tsx:90]: Added extra `Impl` component so we can get access to the `useCollection` hook.

File: packages/react/menu/src/Menu.tsx
Comment (benoitgrelard) [inline packages/react/menu/src/Menu.tsx:70]: Is there a way to get this dynamically from the type of element actually defined in the menu item rather than duplicating here with the risk of getting out of sync?

**Review by benoitgrelard — COMMENTED** [review]
@jjenzz This is now ready for review again. The last commit contains the changes to `@radix-ui/react-collection` to use `@radix-ui/react-slot`.

I've left a few new comments/questions inline for you below.

Also, I'm pretty sure that we'll be able to reuse the collection we've just put in place in `Menu` and its items for the typeahead stuff instead of the custom hook. I think it'll help also to bring it inside the main file as part of the process, all in one place.

File: packages/react/menu/src/Menu.tsx
Comment (jjenzz) [reply packages/react/menu/src/Menu.tsx:70]: Do you mean `React.ElementRef<typeof MenuItem>` ?

File: packages/react/menu/src/Menu.tsx
Comment (jjenzz) [reply packages/react/menu/src/Menu.tsx:300]: `onFocusEnter` isn't `onFocus` in my mind because React `onFocus` is actually `onFocusIn` 🙈  I'm imagining `onMouseEnter` which doesn't bubble, so it is only triggered when mouse first enters the element and not its children. I was applying that same thinking here, so `onFocusEnter` would be when focus first enters the wrapper and not when it's focusing children. Isn't that what this is?

**Review by jjenzz — APPROVED** [review]
I've left a comment above about the `onEntryFocus` naming but just something to consider, I'll leave it up to you as I don't want to hold this up because of it.

File: packages/react/menu/src/Menu.tsx
Comment (benoitgrelard) [reply packages/react/menu/src/Menu.tsx:300]: I get that, but my previous comment was more that I think the event should reflect the specific "what", rather than the "when", seeing as that's the behaviour we want to cancel by "preventDefaulting"ing it.

I know one could say the opposite for example, that on a `select`, arrow down to open isn't an event called `onArrowDownOpen` that you prevent, but you'd prevent the keydown when the key is arrow (ie. it's more about the "when" than the "why") but then if we want to go for that, then I'm not sure why we need a custom event at all, we would just stay on what I had originally which was to `preventDefault` in `onFocus`.

What do you think?

Comment (jjenzz) [reply packages/react/menu/src/Menu.tsx:300]: Interesting. Didn't we move away from `onFocus` because that would mean preventing browser behaviour too? I'm good to stick with `onEntryFocus` :+1:

Comment (benoitgrelard) [reply packages/react/menu/src/Menu.tsx:300]: Well, there isn't anything to prevent natively `onFocus` but the reason we moved away was rather that it seemed like implicit knowledge to have the `Menu` component prevent default in `onFocus` just because we happen to know that the `RovingFocusGroup` is doing something in the `onFocus`.

But yeah I think it's enough bikeshedding here considering that it's not really public API anyway.

**Comment by jjenzz** [comment]
All good from me except this one is still there if that's the answer you were looking for?  https://github.com/radix-ui/primitives/pull/618#discussion_r623906613

File: packages/react/menu/src/Menu.tsx
Comment (benoitgrelard) [reply packages/react/menu/src/Menu.tsx:70]: of course! why do I keep forgetting about this?? 😀

**Pull request merged**
